process/track/blockProcessor.go

```diff
@@ -0,0 +1,265 @@
+	}
+}
+
+func (bp *blockProcessor) computeLongestChainFromLastCrossNotarized(
+	shardID uint32,
+) (data.HeaderHandler, []byte, []data.HeaderHandler, [][]byte) {
+
+	lastCrossNotarizedHeader, lastCrossNotarizedHeaderHash, err := bp.crossNotarizer.getLastNotarizedHeader(shardID)
+	if err != nil {
+		return nil, nil, nil, nil
+	}
+
+	headers, hashes := bp.computeLongestChain(shardID, lastCrossNotarizedHeader)
+	return lastCrossNotarizedHeader, lastCrossNotarizedHeaderHash, headers, hashes
+}
+
+func (bp *blockProcessor) computeSelfNotarizedHeaders(headers []data.HeaderHandler) ([]data.HeaderHandler, [][]byte) {
+	selfNotarizedHeadersInfo := make([]*headerInfo, 0)
+
+	for _, header := range headers {
+		selfHeadersInfo := bp.blockTracker.getSelfHeaders(header)
+		if len(selfHeadersInfo) > 0 {
+			selfNotarizedHeadersInfo = append(selfNotarizedHeadersInfo, selfHeadersInfo...)
+		}
+	}
+
+	if len(selfNotarizedHeadersInfo) > 1 {
+		sort.Slice(selfNotarizedHeadersInfo, func(i, j int) bool {
+			return selfNotarizedHeadersInfo[i].header.GetNonce() < selfNotarizedHeadersInfo[j].header.GetNonce()
+		})
+	}
+
+	selfNotarizedHeaders := make([]data.HeaderHandler, 0)
+	selfNotarizedHeadersHashes := make([][]byte, 0)
+
+	for _, selfNotarizedHeaderInfo := range selfNotarizedHeadersInfo {
+		selfNotarizedHeaders = append(selfNotarizedHeaders, selfNotarizedHeaderInfo.header)
+		selfNotarizedHeadersHashes = append(selfNotarizedHeadersHashes, selfNotarizedHeaderInfo.hash)
+	}
+
+	return selfNotarizedHeaders, selfNotarizedHeadersHashes
+}
+
+func (bp *blockProcessor) computeLongestChain(shardID uint32, header data.HeaderHandler) ([]data.HeaderHandler, [][]byte) {
+	headers := make([]data.HeaderHandler, 0)
+	headersHashes := make([][]byte, 0)
+
+	if check.IfNil(header) {
+		return headers, headersHashes
+	}
+
+	sortedHeaders, sortedHeadersHashes := bp.blockTracker.sortHeadersFromNonce(shardID, header.GetNonce()+1)
+	if len(sortedHeaders) == 0 {
+		return headers, headersHashes
+	}
+
+	longestChainHeadersIndexes := make([]int, 0)
+	headersIndexes := make([]int, 0)
+	bp.getNextHeader(&longestChainHeadersIndexes, headersIndexes, header, sortedHeaders, 0)
+
+	for _, index := range longestChainHeadersIndexes {
+		headers = append(headers, sortedHeaders[index])
+		headersHashes = append(headersHashes, sortedHeadersHashes[index])
+	}
+
+	return headers, headersHashes
+}
+
+func (bp *blockProcessor) getNextHeader(
+	longestChainHeadersIndexes *[]int,
+	headersIndexes []int,
+	prevHeader data.HeaderHandler,
+	sortedHeaders []data.HeaderHandler,
+	index int,
+) {
+	defer func() {
+		if len(headersIndexes) > len(*longestChainHeadersIndexes) {
+			*longestChainHeadersIndexes = headersIndexes
+		}
+	}()
+
+	if check.IfNil(prevHeader) {
+		return
+	}
+
+	for i := index; i < len(sortedHeaders); i++ {
+		currHeader := sortedHeaders[i]
+		if currHeader.GetNonce() > prevHeader.GetNonce()+1 {
+			break
+		}
+
+		if currHeader.GetNonce() == prevHeader.GetNonce()+1 {
```

Comment:
this check exists inside IsHeaderConstructionValid so it is not needed here.

Reply:
Done. I just wanted to avoid the consumed time with the other checks which are made before the nonce check in headerValidator, as here in checkHeaderFinality the headers are already sorted. Now I have checked again the IsHeaderConstructionValid method and it checks only the round before the nonce so I removed this check here.